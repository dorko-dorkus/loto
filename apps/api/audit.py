from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime, timedelta, timezone
import json
import os
import sqlite3
import time
from pathlib import Path

import structlog

DB_PATH = Path(__file__).resolve().parents[2] / "loto.db"


logger = structlog.get_logger(__name__)


def _redact(value: str) -> str:
    """Redact sensitive tokens from ``value``.

    Environment variables such as ``MAXIMO_APIKEY`` may be logged and must be
    removed prior to export.  Common secrets are replaced with ``[REDACTED]``.
    """

    patterns = [
        os.getenv("MAXIMO_APIKEY"),
        os.getenv("MAXIMO_BASE_URL"),
        "apikey",
        "password",
        "secret",
    ]
    redacted = value
    for p in patterns:
        if p:
            redacted = redacted.replace(p, "[REDACTED]")
    return redacted


@dataclass(frozen=True)
class AuditRecord:
    """Represents a single audit log entry."""

    id: int | None
    user: str
    action: str
    timestamp: str


def add_record(*, user: str, action: str, db_path: Path = DB_PATH) -> None:
    """Insert an audit record into the database."""
    conn = sqlite3.connect(db_path)
    try:
        conn.execute(
            "INSERT INTO audit_records (user, action) VALUES (?, ?)",
            (user, action),
        )
        conn.commit()
    finally:
        conn.close()


def export_records(
    bucket: str,
    prefix: str = "audit",
    *,
    db_path: Path = DB_PATH,
    retention_years: int | None = None,
    max_attempts: int = 3,
) -> str:
    """Export audit records to S3 with object lock enabled.

    Parameters
    ----------
    bucket:
        Destination S3 bucket.
    prefix:
        Object key prefix.  Keys are further partitioned by ``YYYY/MM/DD``.
    db_path:
        Path to the SQLite database containing ``audit_records``.
    retention_years:
        Optional number of years to retain the uploaded object.  Defaults to the
        ``AUDIT_RETENTION_YEARS`` environment variable or 7 years if unset.
    max_attempts:
        Number of S3 upload attempts on 5xx errors.

    The uploaded object is retained for the configured number of years to
    satisfy compliance requirements.  Records are uploaded as JSON Lines (JSONL)
    format with one JSON object per line.
    """
    import boto3  # imported lazily to avoid hard dependency during normal use
    from botocore.exceptions import ClientError

    conn = sqlite3.connect(db_path)
    try:
        rows = conn.execute(
            "SELECT id, user, action, timestamp FROM audit_records ORDER BY id"
        ).fetchall()
    finally:
        conn.close()

    body_lines = [
        json.dumps(
            {
                "id": r[0],
                "user": _redact(r[1]),
                "action": _redact(r[2]),
                "timestamp": r[3],
            }
        )
        for r in rows
    ]
    body = ("\n".join(body_lines) + "\n").encode("utf-8") if body_lines else b""

    now = datetime.now(tz=timezone.utc)
    key = f"{prefix}/{now:%Y/%m/%d}/{now.isoformat()}.jsonl"

    retention_years = retention_years or int(os.getenv("AUDIT_RETENTION_YEARS", "7"))
    retain_until = now + timedelta(days=365 * retention_years)

    logger.info(
        "uploading_audit_records",
        bucket=bucket,
        key=key,
        count=len(rows),
        retain_until=retain_until.isoformat(),
    )

    s3 = boto3.client("s3")
    attempt = 0
    while True:
        attempt += 1
        try:
            s3.put_object(
                Bucket=bucket,
                Key=key,
                Body=body,
                ObjectLockMode="COMPLIANCE",
                ObjectLockRetainUntilDate=retain_until,
            )
            break
        except ClientError as exc:  # pragma: no cover - network errors are rare
            status = exc.response.get("ResponseMetadata", {}).get("HTTPStatusCode", 0)
            if 500 <= status < 600 and attempt < max_attempts:
                time.sleep(2 ** (attempt - 1))
                continue
            raise
    return key


if __name__ == "__main__":
    import argparse

    parser = argparse.ArgumentParser(description="Export audit logs to S3")
    parser.add_argument("bucket", help="Destination S3 bucket")
    parser.add_argument("--prefix", default="audit", help="Object prefix")
    args = parser.parse_args()
    export_records(args.bucket, prefix=args.prefix)
