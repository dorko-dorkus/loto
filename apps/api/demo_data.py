from __future__ import annotations

import json
from pathlib import Path
from typing import Any, Dict, List

DATA_DIR = Path(__file__).with_name("demo_data")


class DemoDataSource:
    """Load demo data from JSON files on startup."""

    def __init__(self) -> None:
        self.work_orders: List[Dict[str, Any]] = self._load_list("workorders.json")
        self.assets: List[Dict[str, Any]] = self._load_list("assets.json")
        self.locations: List[Dict[str, Any]] = self._load_list("locations.json")
        self.inventory: List[Dict[str, Any]] = self._load_list("inventory.json")
        self.blueprints: Dict[str, Dict[str, Any]] = self._load_dict("blueprints.json")
        self.boms: List[Dict[str, Any]] = self._load_list("bom.json")
        self._work_orders_by_id = {wo["id"]: wo for wo in self.work_orders}
        self._bom_by_wo: Dict[str, List[Dict[str, Any]]] = {}
        for line in self.boms:
            self._bom_by_wo.setdefault(line["workorder"], []).append(line)
        self.asset_ids = {asset["id"] for asset in self.assets}
        self.location_ids = {loc["id"] for loc in self.locations}

    def _load_list(self, filename: str) -> List[Dict[str, Any]]:
        path = DATA_DIR / filename
        if path.exists():
            return json.loads(path.read_text())
        return []

    def _load_dict(self, filename: str) -> Dict[str, Dict[str, Any]]:
        path = DATA_DIR / filename
        if path.exists():
            return json.loads(path.read_text())
        return {}

    def list_work_orders(self) -> List[Dict[str, Any]]:
        return list(self._work_orders_by_id.values())

    def get_work_order(self, work_order_id: str) -> Dict[str, Any]:
        return self._work_orders_by_id[work_order_id]

    def get_blueprint(self, work_order_id: str) -> Dict[str, Any] | None:
        return self.blueprints.get(work_order_id)

    def get_bom(self, work_order_id: str) -> List[Dict[str, Any]]:
        """Return bill-of-material lines for ``work_order_id``."""

        return list(self._bom_by_wo.get(work_order_id, []))

    def validate(self) -> Dict[str, List[Dict[str, str | None]]]:
        """Check referential integrity of work orders.

        Returns a mapping with lists of missing assets and locations.
        """
        missing_assets: List[Dict[str, str | None]] = []
        missing_locations: List[Dict[str, str | None]] = []
        for wo in self.work_orders:
            asset = wo.get("assetnum")
            if not asset or asset not in self.asset_ids:
                missing_assets.append(
                    {"workorder": wo.get("id", ""), "assetnum": asset}
                )
            location = wo.get("location")
            if not location or location not in self.location_ids:
                missing_locations.append(
                    {"workorder": wo.get("id", ""), "location": location}
                )
        return {
            "missing_assets": missing_assets,
            "missing_locations": missing_locations,
        }


demo_data = DemoDataSource()
