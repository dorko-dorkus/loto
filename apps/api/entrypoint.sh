#!/bin/sh
set -e
alembic -c apps/api/alembic/alembic.ini upgrade head
exec uvicorn apps.api.main:app --host 0.0.0.0 --port 8000
