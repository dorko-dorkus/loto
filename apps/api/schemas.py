from __future__ import annotations

from datetime import datetime
from typing import Dict, List

from pydantic import BaseModel, Field


class BlueprintRequest(BaseModel):
    """Request body for the /blueprint endpoint."""

    workorder_id: str = Field(..., description="Identifier of the work order")

    class Config:
        extra = "forbid"


class Step(BaseModel):
    """Single isolation step in the computed plan."""

    component_id: str = Field(..., description="Component identifier")
    method: str = Field(..., description="Isolation method")

    class Config:
        extra = "forbid"


class BlueprintResponse(BaseModel):
    """Response model for the /blueprint endpoint."""

    steps: List[Step] = Field(
        default_factory=list, description="Ordered isolation steps"
    )
    unavailable_assets: List[str] = Field(
        default_factory=list, description="Assets that became unavailable"
    )
    unit_mw_delta: Dict[str, float] = Field(
        default_factory=dict, description="Lost capacity per unit in MW"
    )
    blocked_by_parts: bool = Field(
        False, description="Whether execution is blocked due to missing parts"
    )
    parts_status: Dict[str, str] = Field(
        default_factory=dict, description="Inventory status per material line"
    )

    class Config:
        extra = "forbid"


class ScheduleRequest(BaseModel):
    """Request body for the /schedule endpoint."""

    workorder: str = Field(..., description="Identifier of the work order")

    class Config:
        extra = "forbid"


class SchedulePoint(BaseModel):
    """Single datapoint in the returned schedule."""

    date: str = Field(..., description="ISO formatted date")
    p10: float = Field(..., description="P10 duration percentile")
    p50: float = Field(..., description="P50 duration percentile")
    p90: float = Field(..., description="P90 duration percentile")
    price: float = Field(..., description="Energy price for the interval")
    hats: int = Field(..., description="Crew size (number of hard hats)")

    class Config:
        extra = "forbid"


class ScheduleResponse(BaseModel):
    """Response model for the /schedule endpoint."""

    schedule: List[SchedulePoint] = Field(
        default_factory=list, description="Synthetic schedule data"
    )
    seed: str = Field(..., description="Seed used for deterministic simulation")
    objective: float = Field(..., description="Objective value for the schedule")
    blocked_by_parts: bool = Field(
        False,
        description="Whether execution is blocked due to missing parts",
    )
    rulepack_sha256: str = Field(
        ..., description="SHA-256 digest of the rule pack used"
    )
    rulepack_id: str | None = Field(None, description="Identifier of the rule pack")
    rulepack_version: str | None = Field(None, description="Version of the rule pack")

    class Config:
        extra = "forbid"


class HatKpiRequest(BaseModel):
    """Payload for recording hat KPI metrics."""

    wo_id: str = Field(..., description="Work order identifier")
    hat_id: str = Field(..., description="Hat identifier")
    SA: float = Field(..., description="Safety metric A")
    SP: float = Field(..., description="Safety metric P")
    RQ: float | None = Field(None, description="Optional RQ metric")
    OF: float | None = Field(None, description="Optional OF metric")

    class Config:
        extra = "forbid"


class HatSnapshot(BaseModel):
    """Snapshot of ranking information for a hat."""

    hat_id: str = Field(..., description="Identifier of the hat")
    rank: int = Field(0, description="Rank among hats (1 = best)")
    c_r: float = Field(0.5, description="Ranking coefficient")
    n_samples: int = Field(0, description="Number of KPI events")
    last_event_at: datetime | None = Field(
        None,
        description="Timestamp of the most recent event",
    )

    class Config:
        extra = "forbid"
