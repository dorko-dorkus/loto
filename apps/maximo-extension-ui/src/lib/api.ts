/**
 * Helper to prefix API requests with the configured base URL.
 *
 * Uses the `NEXT_PUBLIC_API_BASE` environment variable if set.
 */
export function apiFetch(path: string, init?: RequestInit): Promise<Response> {
  const base = process.env.NEXT_PUBLIC_API_BASE ?? '';
  return fetch(`${base}${path}`, init);
}
