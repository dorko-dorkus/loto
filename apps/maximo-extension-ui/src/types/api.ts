export interface WorkOrderSummary {
  id: string;
  description: string;
  status: string;
  owner?: string;
  plannedStart?: string;
  plannedFinish?: string;
}

export interface BlueprintStep {
  component_id: string;
  method: string;
}

export interface BlueprintData {
  steps: BlueprintStep[];
  unavailable_assets: string[];
  unit_mw_delta: Record<string, number>;
  blocked_by_parts?: boolean;
  parts_status?: Record<string, MaterialStatus>;
  diff?: string;
  audit_metadata?: Record<string, unknown>;
}

export interface PlanStep {
  step: number;
  description: string;
  resources: string;
}

export interface SimulationResult {
  id: number;
  deltaTime: string;
  deltaCost: string;
  readiness: string;
  isolation: boolean;
}

export interface ImpactRecord {
  metric: string;
  before: number;
  after: number;
  delta: number;
}

export type InventoryStatus = 'ready' | 'short' | 'ordered';

export type MaterialStatus = 'ok' | 'low' | 'short' | 'rfq' | 'parked';

export interface InventoryItem {
  item: string;
  required: number;
  onHand: number;
  eta?: string;
  status: InventoryStatus;
}

export interface HatSnapshot {
  hat_id: string;
  rank: number;
  c_r: number;
  n_samples: number;
  last_event_at?: string;
}
