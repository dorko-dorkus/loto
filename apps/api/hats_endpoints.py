from __future__ import annotations

import os
from collections import defaultdict
from pathlib import Path
from typing import Dict, List

from fastapi import APIRouter, HTTPException

from loto.roster import storage, update_ranking

from .schemas import HatKpiRequest, HatKpiResponse

router = APIRouter(prefix="/hats", tags=["hats"])


def _ledger_path() -> Path:
    return Path(os.getenv("HATS_LEDGER_FILE", "hats_ledger.jsonl"))


def _snapshot_path() -> Path:
    return Path(os.getenv("HATS_SNAPSHOT_FILE", "hats_snapshot.json"))


@router.post("/kpi", response_model=HatKpiResponse)
async def post_kpi(payload: HatKpiRequest) -> HatKpiResponse:
    """Record KPI metrics for a hat and return updated ranking."""

    entry = payload.dict(exclude_none=True)
    ledger_path = _ledger_path()
    snapshot_path = _snapshot_path()
    try:
        storage.append_ledger(ledger_path, entry)
    except ValueError:
        # Duplicate entry â€“ idempotent behaviour
        pass

    entries = storage.read_ledger(ledger_path)
    snapshot = storage.compute_snapshot(entries)
    storage.write_snapshot(snapshot_path, snapshot)

    ledger: Dict[str, List[List[float]]] = defaultdict(list)
    for e in snapshot.values():
        metrics: List[float] = []
        for key in ("SA", "SP", "RQ", "OF"):
            if key in e:
                metrics.append(float(e[key]))
        ledger[str(e["hat_id"])].append(metrics)

    ranking = update_ranking(ledger)
    data = ranking.get(payload.hat_id)
    if data is None:
        raise HTTPException(status_code=500, detail="ranking failed")
    return HatKpiResponse(**data)
