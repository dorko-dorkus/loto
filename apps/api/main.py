from __future__ import annotations

import logging
import os
import time
from dataclasses import dataclass
from datetime import date, timedelta
from pathlib import Path
from typing import Any, Dict, List
from uuid import uuid4

from fastapi import FastAPI, HTTPException, Request, Response
from fastapi.middleware.cors import CORSMiddleware
from fastapi.responses import JSONResponse
from pydantic import BaseModel, Field

from loto.impact_config import load_impact_config
from loto.integrations.stores_adapter import DemoStoresAdapter
from loto.inventory import (
    InventoryStatus,
    Reservation,
    StockItem,
    check_wo_parts_required,
)
from loto.loggers import configure_logging, request_id_var, rule_hash_var, seed_var
from loto.materials.jobpack import DEFAULT_LEAD_DAYS, build_jobpack
from loto.models import RulePack
from loto.scheduling.des_engine import Task
from loto.scheduling.monte_carlo import simulate
from loto.service import plan_and_evaluate
from loto.service.blueprints import inventory_state

from .pid_endpoints import router as pid_router
from .schemas import (
    BlueprintRequest,
    BlueprintResponse,
    SchedulePoint,
    ScheduleRequest,
    ScheduleResponse,
    Step,
)

configure_logging()

app = FastAPI(title="loto API")

origins = [o.strip() for o in os.getenv("CORS_ORIGINS", "").split(",") if o.strip()]
if origins:
    app.add_middleware(
        CORSMiddleware,
        allow_origins=origins,
        allow_credentials=True,
        allow_methods=["*"],
        allow_headers=["*"],
    )

app.include_router(pid_router)


AUTH_REQUIRED = os.getenv("AUTH_REQUIRED", "").lower() == "true"
AUTH_TOKEN = os.getenv("AUTH_TOKEN", "")


@app.middleware("http")
async def auth_guard(request: Request, call_next):
    if AUTH_REQUIRED:
        auth_header = request.headers.get("Authorization")
        if auth_header != f"Bearer {AUTH_TOKEN}":
            return Response(status_code=401)
    return await call_next(request)


@app.middleware("http")
async def log_context(request: Request, call_next):
    req_id = str(uuid4())
    token = request_id_var.set(req_id)
    try:
        response = await call_next(request)
        return response
    finally:
        request_id_var.reset(token)
        seed_var.set(None)
        rule_hash_var.set(None)


@app.exception_handler(HTTPException)
async def _handle_http_exception(request: Request, exc: HTTPException) -> JSONResponse:
    """Return errors in a consistent JSON envelope."""
    return JSONResponse(status_code=exc.status_code, content={"error": str(exc.detail)})


@app.exception_handler(Exception)
async def _handle_exception(request: Request, exc: Exception) -> JSONResponse:
    """Catch-all handler to wrap unexpected exceptions."""
    logging.exception("Unhandled exception: %s", exc)
    return JSONResponse(status_code=500, content={"error": str(exc)})


ENV = os.getenv("APP_ENV", "").lower()
if ENV == "live":
    ENV_BADGE = "PROD"
elif ENV == "test":
    ENV_BADGE = "TEST"
else:
    ENV_BADGE = "DRY-RUN"

RATE_LIMIT_PATHS = {"/pid/overlay", "/schedule"}
RATE_LIMIT_CAPACITY = int(os.getenv("RATE_LIMIT_CAPACITY", "10"))
RATE_LIMIT_INTERVAL = float(os.getenv("RATE_LIMIT_INTERVAL", "60"))
_rate_limit_state = {
    path: {"tokens": RATE_LIMIT_CAPACITY, "ts": time.monotonic()}
    for path in RATE_LIMIT_PATHS
}


@app.middleware("http")
async def add_env_and_rate_limit(request: Request, call_next):
    path = request.url.path
    if path in _rate_limit_state:
        bucket = _rate_limit_state[path]
        now = time.monotonic()
        elapsed = now - bucket["ts"]
        if elapsed > RATE_LIMIT_INTERVAL:
            bucket["tokens"] = RATE_LIMIT_CAPACITY
            bucket["ts"] = now
        if bucket["tokens"] <= 0:
            response = Response(status_code=429)
            response.headers["X-Env"] = ENV_BADGE
            return response
        bucket["tokens"] -= 1
    response = await call_next(request)
    response.headers["X-Env"] = ENV_BADGE
    return response


STATE: Dict[str, Any] = {}


@dataclass
class WorkOrder:
    """Minimal work order representation for demo inventory checks."""

    id: str
    reservations: List[Reservation]


class ProposeRequest(BaseModel):
    """Payload for proposing plan and schedule updates."""

    plan: Dict[str, Any] = Field(default_factory=dict, description="Proposed targets")
    schedule: Dict[str, Any] = Field(
        default_factory=dict, description="Proposed assignments"
    )

    class Config:
        extra = "forbid"


class ProposeResponse(BaseModel):
    """Response highlighting differences from current state."""

    diff: Dict[str, Dict[str, Dict[str, Any]]] = Field(
        default_factory=dict, description="Differences vs current targets/assignments"
    )
    idempotency_key: str

    class Config:
        extra = "forbid"


@app.get("/healthz", include_in_schema=False)
async def healthz() -> dict[str, Any]:
    """Health check endpoint including rate limit counters."""
    return {
        "status": "ok",
        "rate_limit": {
            "capacity": RATE_LIMIT_CAPACITY,
            "interval": RATE_LIMIT_INTERVAL,
            "counters": {
                path: state["tokens"] for path, state in _rate_limit_state.items()
            },
        },
    }


class DemoMaximoAdapter:
    """Tiny Maximo adapter serving demo data from the repository."""

    def load_context(self, workorder_id: str) -> Dict[str, Any]:
        """Return file paths and metadata for the given work order."""

        base = Path(__file__).resolve().parents[2] / "demo"
        impact_cfg = load_impact_config(
            base / "unit_map.yaml", base / "redundancy_map.yaml"
        )
        return {
            "line_csv": base / "line_list.csv",
            "valve_csv": base / "valves.csv",
            "drain_csv": base / "drains.csv",
            "source_csv": base / "sources.csv",
            "asset_tag": "A",  # arbitrary asset tag for demo data
            "impact_cfg": impact_cfg,
        }


@app.post("/blueprint", response_model=BlueprintResponse)
async def post_blueprint(payload: BlueprintRequest) -> BlueprintResponse:
    """Plan isolations for a work order and return impact metrics."""

    adapter = DemoMaximoAdapter()
    ctx = adapter.load_context(payload.workorder_id)
    impact_cfg = ctx["impact_cfg"]

    stores = DemoStoresAdapter()
    work_order = WorkOrder(
        id=payload.workorder_id,
        reservations=[
            Reservation(item_id="P-100", quantity=1),
            Reservation(item_id="P-200", quantity=1),
        ],
    )

    def lookup_stock(item_id: str) -> StockItem | None:
        try:
            status = stores.inventory_status(item_id)
        except KeyError:
            return None
        return StockItem(item_id=item_id, quantity=status.get("available", 0))

    def check_parts(wo: object) -> InventoryStatus:
        assert isinstance(wo, WorkOrder)
        return check_wo_parts_required(wo, lookup_stock)

    parts_status: Dict[str, str] = {}
    for res in work_order.reservations:
        stock = lookup_stock(res.item_id)
        available = stock.quantity if stock else 0
        if available >= res.quantity:
            parts_status[res.item_id] = "ok"
        elif available > 0:
            parts_status[res.item_id] = "low"
        else:
            parts_status[res.item_id] = "short"

    inv_status = check_parts(work_order)

    global STATE
    STATE = dict(inventory_state(work_order, check_parts, STATE))

    with (
        open(ctx["line_csv"]) as line,
        open(ctx["valve_csv"]) as valve,
        open(ctx["drain_csv"]) as drain,
        open(ctx["source_csv"]) as source,
    ):
        plan, _, impact, prov = plan_and_evaluate(
            line,
            valve,
            drain,
            source,
            asset_tag=str(ctx["asset_tag"]),
            rule_pack=RulePack(),
            stimuli=[],
            asset_units=impact_cfg.asset_units,
            unit_data=impact_cfg.unit_data,
            unit_areas=impact_cfg.unit_areas,
            penalties=impact_cfg.penalties,
            asset_areas=impact_cfg.asset_areas,
        )
    seed_var.set(prov.seed)
    rule_hash_var.set(prov.rule_hash)
    logging.info("request complete")

    steps: List[Step] = [
        Step(component_id=a.component_id, method=a.method) for a in plan.actions
    ]

    return BlueprintResponse(
        steps=steps,
        unavailable_assets=sorted(impact.unavailable_assets),
        unit_mw_delta=impact.unit_mw_delta,
        blocked_by_parts=inv_status.blocked,
        parts_status=parts_status,
    )


def _diff(
    current: Dict[str, Any], proposed: Dict[str, Any]
) -> Dict[str, Dict[str, Any]]:
    """Return key-wise difference between two mappings."""

    delta: Dict[str, Dict[str, Any]] = {}
    for key in set(current) | set(proposed):
        cur = current.get(key)
        prop = proposed.get(key)
        if cur != prop:
            delta[key] = {"current": cur, "proposed": prop}
    return delta


@app.post("/propose", response_model=ProposeResponse)
async def post_propose(payload: ProposeRequest) -> ProposeResponse:
    """Return diffs between proposed plan/schedule and current state."""

    current: Dict[str, Dict[str, Any]] = {
        "targets": {"A": 1},
        "assignments": {"A": "crew-1"},
    }

    diff = {
        "targets": _diff(current["targets"], payload.plan),
        "assignments": _diff(current["assignments"], payload.schedule),
    }

    return ProposeResponse(diff=diff, idempotency_key=str(uuid4()))


@app.post("/schedule", response_model=ScheduleResponse)
async def post_schedule(payload: ScheduleRequest) -> ScheduleResponse:
    """Return a synthetic schedule for the given work order."""

    # Minimal demo task graph
    tasks = {
        "prep": Task(duration=1),
        "exec": Task(duration=2, predecessors=["prep"]),
    }

    mc_res = simulate(tasks, {}, runs=20)
    today = date.today()
    schedule: List[SchedulePoint] = []
    for i, pct in enumerate(mc_res.task_percentiles.values()):
        schedule.append(
            SchedulePoint(
                date=(today + timedelta(days=i)).isoformat(),
                p10=pct.get("P10", 0.0),
                p50=pct.get("P50", 0.0),
                p90=pct.get("P90", 0.0),
                price=0.0,
                hats=i + 1,
            )
        )

    seed_int = 0
    seed_var.set(seed_int)
    logging.info("request complete")
    return ScheduleResponse(schedule=schedule, seed=str(seed_int), objective=0.0)


@app.get("/workorders/{workorder_id}")
async def get_workorder(workorder_id: str) -> dict[str, str]:
    """Mock work order fetch."""
    return {"workorder_id": workorder_id, "status": "mocked"}


@app.get("/workorders/{workorder_id}/jobpack")
async def get_jobpack(
    workorder_id: str,
    permit_start: date | None = None,
    lead_days: int = DEFAULT_LEAD_DAYS,
) -> dict[str, dict[str, object]]:
    """Return a mock job pack for the given work order."""

    return build_jobpack(workorder_id, permit_start=permit_start, lead_days=lead_days)
