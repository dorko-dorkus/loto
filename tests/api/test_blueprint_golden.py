from __future__ import annotations

import json
from pathlib import Path

import pytest
from fastapi.testclient import TestClient

from apps.api.main import app
from apps.api.schemas import BlueprintResponse


@pytest.fixture
def golden(request):
    path = Path(request.node.fspath).with_suffix(".golden.json")
    expected = json.loads(path.read_text())

    def check(data: object) -> None:
        assert data == expected

    return check


def test_blueprint_golden(golden):
    client = TestClient(app)
    payload = {"workorder_id": "WO-1"}
    res1 = client.post("/blueprint", json=payload)
    assert res1.status_code == 200
    data1 = BlueprintResponse.model_validate(res1.json()).model_dump()
    golden(data1)

    res2 = client.post("/blueprint", json=payload)
    assert res2.status_code == 200
    data2 = BlueprintResponse.model_validate(res2.json()).model_dump()
    assert data2 == data1
