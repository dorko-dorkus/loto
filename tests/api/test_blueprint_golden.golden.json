{
  "blocked_by_parts": false,
  "parts_status": {
    "P-100": "ok",
    "P-200": "ok"
  },
  "steps": [
    {
      "component_id": "VALVE-1",
      "method": "lock"
    }
  ],
  "unavailable_assets": [],
  "unit_mw_delta": {}
}